src/blockchain/compact_filters/p2p.rs

```diff
@@ -0,0 +1,646 @@
+            threads,
+            cbf_buffer: min_cbf,
+            non_cbf_buffer: min_non_cbf,
+            progress,
+        }
+    }
+
+    /// Get running address discovery progress
+    fn get_progress(&self) -> DiscoveryData {
+        let (queued_count, visited_count) = {
+            let address_discovery = self.discovery.read().unwrap();
+            (
+                address_discovery.pending.len(),
+                address_discovery.visited.len(),
+            )
+        };
+
+        let cbf_node_count = self.directory.get_cbf_address_count();
+        let other_node_count = self.directory.get_non_cbf_address_count();
+
+        DiscoveryData {
+            queued: queued_count,
+            visited: visited_count,
+            non_cbf_count: cbf_node_count + other_node_count,
+            cbf_count: cbf_node_count,
+        }
+    }
+
+    /// Spawn [self.thread] no. of worker threads
+    fn spawn_workers(&mut self) -> Vec<JoinHandle<()>> {
+        let mut worker_handles: Vec<JoinHandle<()>> = vec![];
+        for _ in 0..self.threads {
+            let sender = self.sender.clone();
+            let discovery = self.discovery.clone();
+            let network = self.network;
+            let worker_handle = thread::spawn(move || {
+                let mut worker = AddressWorker::new(discovery, sender, network);
+                worker.work();
+            });
+            worker_handles.push(worker_handle);
+        }
+        worker_handles
+    }
+
+    /// Crawl the Bitcoin network until given number of cbf/non_cbf nodes are found
+    ///
+    /// - This will start a bunch of crawlers.
+    /// - load up the existing cache.
+    /// - Update the cache with new found peers.
+    /// - check if address is in banlist
+    /// - run crawlers until buffer requirement is matched
+    /// - flush the current cache into disk
+    pub fn fetch(&mut self) {
+        self.spawn_workers();
+
+        // Get already existing cache
+        let mut cache = match AddressCache::from_file(&self.cache_filename) {
+            Some(cache) => cache,
+            None => AddressCache::empty(),
+        };
+
+        while self.directory.get_cbf_buffer() < self.cbf_buffer
+            || self.directory.get_non_cbf_buffer() < self.non_cbf_buffer
+        {
+            if let Ok(message) = self.receiver.recv() {
+                match message {
+                    AddressMessage::NewValid(addr, flag) => {
+                        if !cache.banned_peers.contains(&addr) {
+                            let cbf = flag.has(ServiceFlags::COMPACT_FILTERS);
+
+                            self.directory.add_address(addr, cbf);
+
+                            cache.add_address(addr, cbf);
+                        }
+                    }
+                }
+            }
+            self.progress.update(self.get_progress()).unwrap();
+        }
+
+        // When completed, flush the cache
+        cache.write_to_file(&self.cache_filename).unwrap();
```

Comment:
Maybe it makes sense to write the cache periodically and not just at the end?

Reply:
Yes it does. But that involves threading. I thought of that first as it made more sense, but having multiple threads made stuff difficult. So I ended up writing everything single-threaded and optimized around blockings. 

As mentioned in https://github.com/bitcoindevkit/bdk/pull/378#discussion_r660353392 we can expect `fetch` to be called occasionally. As a result, cache will also update occasionally, and we will know exactly when it updates, only in a `fetch` call. 

Thus it made sense to not give its own dedicated loop and just use file i/o only when needed.

Reply:
At one point I had it writing the cache to file inside of the fetch() message loop from workers anytime I updated the memory cache.  I ran into a 'too many open files' issue though and so moved it out.  Could probably get away with some kind or probabilistic flush if we want to update it more often but not every iteration of the loop.

Reply:
Yes, That's the reason I took it out of the loop. So currently it's not flushing at each iteration. It flushes after the loop completes. So that problem doesn't exist now.

https://github.com/bitcoindevkit/bdk/blob/23fe55127e0a4c6871df0ff66ce8897aeff78143/src/blockchain/compact_filters/address_manager.rs#L539

Also, note that the loop will not start in case the buffer requirement is already fulfilled. Flush will happen only once per `fetch()` or `ban()` call. Because these are the only calls where we expect the cache to be updated.